web/src/typeahead_helper.ts

```diff
@@ -222,93 +223,194 @@ export function sorter<T>(query: string, objs: T[], get_item: (x: T) => string):
-    if (compose_state.get_message_type() !== "private") {
-        if (person_a.type === "broadcast") {
-            if (person_b.type === "broadcast") {
-                return person_a.user.idx - person_b.user.idx;
-            }
-            return -1;
-        } else if (person_b.type === "broadcast") {
-            return 1;
+    if (person_a.type === "user") {
+        if (person_b.type === "user") {
+            // If both are users
+            return compare_users_for_streams(person_a.user, person_b.user, stream_id, topic);
         }
-    } else {
-        if (person_a.type === "broadcast") {
-            if (person_b.type === "broadcast") {
-                return person_a.user.idx - person_b.user.idx;
-            }
-            return 1;
-        } else if (person_b.type === "broadcast") {
-            return -1;
+        // a is user, b is wildcard
+        const a_is_sub = stream_data.is_user_subscribed(stream_id, person_a.user.user_id);
+        const a_is_sender_to_current_topic =
+            recent_senders.max_id_for_stream_topic_sender({
+                stream_id,
+                topic,
+                sender_id: person_a.user.user_id,
+            }) > 0;
+        // The max_id_for_stream_topic_sender() will be > 0 if the user have sent at least one message to the topic
+        const a_is_sender_to_current_stream =
+            recent_senders.max_id_for_stream_sender({
+                stream_id,
+                sender_id: person_a.user.user_id,
+            }) > 0;
+        const a_is_partner = pm_conversations.is_partner(person_a.user.user_id);
+
+        // Wildcard gets higher priority over (In Stream conversations):
+        // The non-subscribers who have neither participated in the topic nor stream
+        // and also with whom the current user hasn't PMed.
+        const properties_to_satisfy = [
+            a_is_sub,
+            a_is_sender_to_current_topic,
+            a_is_sender_to_current_stream,
+            a_is_partner,
+        ];
+        return compare_user_wildcard(properties_to_satisfy, stream_id);
+    }
+    if (person_b.type === "user") {
+        // a is wildcard, b is user
+        const b_is_sub = stream_data.is_user_subscribed(stream_id, person_b.user.user_id);
+        const b_is_sender_to_current_topic =
+            recent_senders.max_id_for_stream_topic_sender({
+                stream_id,
+                topic,
+                sender_id: person_b.user.user_id,
+            }) > 0;
+        const b_is_sender_to_current_stream =
+            recent_senders.max_id_for_stream_sender({
+                stream_id,
+                sender_id: person_b.user.user_id,
+            }) > 0;
+        const b_is_partner = pm_conversations.is_partner(person_b.user.user_id);
+
+        const properties_to_satisfy = [
+            b_is_sub,
+            b_is_sender_to_current_topic,
+            b_is_sender_to_current_stream,
+            b_is_partner,
+        ];
```

Comment:
Oh yeahhh! I have got it. Thanks!

Reply:
Was this feedback addressed? I still `properties_to_satisfy being calculated twice.

Reply:
I think so?
I have moved the implementation to `get_properties_to_satisy` and it is getting called once for a user without repeating the logic.

Am I missing something here?

Reply:
I have already extracted that logic into `get_propeties_to_satisfy` function (starting from line  300).
This previous code from line 340 doesn't exist anymore and is outdated.

Can you see if you are seeing my newly pushed code please? Because I have changed the code into exactly the same thing you have said (calling the function for person_a and person_b -> Lines 361, 368, 387, 391 in the new code).

Reply:
I see, you added the changes to the wrong commit.